README.md
# Hornbill API Scheduler

The utility provides a quick and easy way to schedule the running of Hornbill API's.

## Installation

* Download the OS-specific ZIP archive containing the executables, configuration file and license;
* Extract the ZIP archive into a folder you would like the application to run from e.g. 'C:\hornbill_scheduler\'.

## Configuration

Example JSON File:

```json
{
  "APIKey": "apikeygoeshere",
  "InstanceID": "instanceIDgoeshere",
  "Schedule": [
    {
      "Enabled":true,
      "CronSchedule":"0 1 23 * * 1-6",
      "DayOfMonthANDDayOfWeek": false,
      "ScheduleFrom":"2016-11-12T00:00:00.000Z",
      "ScheduleTo":"2017-01-01T00:00:00.000Z",
      "Service":"apps/com.hornbill.servicemanager/Incidents",
      "API":"logIncident",
      "APIParams":{
        "0":
        {
          "Type":"Content",
          "Parameter":"summary",
          "Content":"Request Summary"
        },
        "1":
        {
          "Type":"Content",
          "Parameter":"description",
          "Content":"Request Description"
        },
        "2":
        {
          "Type":"Content",
          "Parameter":"customerId",
          "Content":"alanc"
        }
      }
    },
    {
      "Enabled":true,
      "CronSchedule":"* * * * * 1-5",
      "DayOfMonthANDDayOfWeek": false,
      "ScheduleFrom":"2016-11-12T00:00:00.000Z",
      "ScheduleTo":"2017-01-01T00:00:00.000Z",
      "Service":"apps/com.hornbill.servicemanager/Requests",
      "API":"updateReqTimeline",
      "APIParams":{
        "0":
        {
          "Type":"Content",
          "Parameter":"requestId",
          "Content":"IN0000018"
        },
        "1":
        {
          "Type":"Content",
          "Parameter":"content",
          "Content":"This is an auto update"
        }
      }
    }
  ]
}
```

* "APIKey" - a Hornbill API key for a user account with the correct permissions to carry out all of the required API calls
* "InstanceId" - the ID of your Hornbill instance
* "Schedule" - A JSON array, where each object within this array contains the configuration for one scheduled and repeatable task:
  * "Enabled" - set to true to enable the schedule item
  * "CronSchedule" - A Cron compatible schedule expression to schedule the API call by
  * "DayOfMonthANDDayOfWeek" - Boolean true or false. When true, the content of BOTH Day of Week and Day of Month parts of the expression will be enforced, rather than the crontab standard of either.
    * NOTE: the following special characters are supported in these parts of the expression when this is set to true `* , - ?`
  * "ScheduleFrom" - An RFC3339 formatted time string, to specify the date & time to start running any instances of the particular schedule entry. This can contain an empty string to allow you to not specify a date/time to start the schedule from
  * "ScheduleTo" - An RFC3339 formatted time string, to specify the date & time to stop running any more instances of the particular schedule entry. This can contain an empty string, should you wish the schedule to run indefinitely
  * "Service" - The Hornbill Service that contains the API you wish to running
  * "API" - The name of the API to run
  * "APIParams" - A JSON object, containing one or more other JSON objects, allowing you to specify the Order that the parameter should be presented to the API, the parameter type, the parameter ID and the content to write:
  * "Type" - Can be set to:
    * 'Content' - To write a parameter name (Parameter) and value (Content);
    * 'Open' - Allows for complex parameters to be written to the API, within the element specified in this node - must be matched with a parameter type of 'Close'
    * 'Close' - Allows for complex parameters to be written to the API, within the element specified in this node - must be matched with a parameter type of 'Open'
  * 'Parameter' - The name of the parameter
  * 'Content' - the string that should be written within the Parameter node.

For API parameters in Hornbill that require a Date/Time string value, rather than a hard-coding this date/time within the configuration, you can specify an expression to write a date/time string whose value is the number of minutes/hours/days/months/years AFTER the date/time that the scheduled event is ran. This can be particularly useful when scheduling the raising of tasks within Hornbill, and you need to specify a targeted date of completion for the task.

The expression should be written in this format:

nowPlus::X::Y

Where X is an integer value, and Y is the unit of time. So for example:

* nowPlus::2::minutes - would return the scheduled time plus 2 minutes
* nowPlus::4::days - would return the scheduled time plus 4 days
* nowPlus::1::years - would return the scheduled time plus 1 year

The units of time currently supported are:

* minutes
* hours
* days
* months
* years

## Execute

Command Line Parameters

* file : This should point to your json configuration file and by default looks for a file in the current working directory called `conf.json`. If this is present you don't need to have the parameter.
* dryrun : Defaults to `false`. Set to `true` to run the tool in dryrun mode, which outputs all API call details (service, method and payload) to the log without actually firing the API calls
* debug : Defaults to `false`. Set to `true` to run the tool in debug mode, which outputs the API call request and reponse XML payload to the log file when not in dryrun mode

`goAPIScheduler.exe -file=conf.json -dryrun=true`

When you are ready to start the scheduler:

* Open '''conf.json''' and add in the necessary configration;
* Open Command Line Prompt as Administrator;
* Change Directory to the folder containing the scheduler executables and configuration file 'C:\hornbill_scheduler\';
* Run the command relevant to your OS: goAPIScheduler.exe Windows systems
* Follow all on-screen prompts, taking careful note of all prompts and messages provided.

When the scheduler is executed, you will be presented with a list of all active schedule items from the configuration file, and these items will be executed as per the config at the relevant dates/times.

## Exit

To end the scheduler app, press CTRL+C in the Command Prompt window where the scheduler is running.


CHANGELOG.md
# CHANGELOG

## 1.3.1 (July 6th, 2021)

Change:

- Rebuilt using latest version of goApiLib, to fix possible issue witt connections via a proxy

## 1.3.0 (March 27th 2019)

Features:

- Added support to allow Day Of Month and Day Of Week parts of expression to both be enforced
- Additional parsing of the cron schedules, better logging when incorrect expression provided

## 1.2.1 (March 12th 2019)

Defect Fix:

- Duplicate runs of scheduled APIs on certain builds of Windows

## 1.2.0 (January 21st, 2019)

Features:

- Added "debug" mode, to allow scheduler to output API call payloads to the log file

## 1.1.0 (December 13th, 2018)

Features:

- Added "dryrun" mode, to allow scheduler to output API call details out to the log without actually firing the API calls
- Removed need to specify instance zone in configuration
- General tidy up of the code

## 1.0.3 (July 4th, 2017)

Update:

- Updated third-party cron library to fix duplicate API call issue.

## 1.0.2 (February 7th, 2017)

Defect fix:

- Unable to use non-default configuration file name

## 1.0.1 (November 14th, 2016)

Features:

- Added ability to generate dynamic date strings for insertion in to API parameters

## 1.0.0 (November 12th, 2016)

Features:

- Initial Release

main.structs.go
package main

const (
	version       = "1.3.1"
	timeLayout    = "2006-01-02T15:04:05.000Z"
	cliTimeLayout = "2006/01/02 15:04:05"
)

var (
	configFileName string
	configDebug    bool
	configDryRun   bool
	configVersion  bool
	logEntryType   = 3
	apiCallConfig  apiCallStruct
	timeNow        string
	dowMap         = make(map[string]int)
)

//----- Config Data Structs
type apiCallStruct struct {
	APIKey     string
	InstanceID string
	Zone       string
	Schedule   []apiSchedStruct
}

type apiSchedStruct struct {
	Enabled                bool
	CronSchedule           string
	DayOfMonthANDDayOfWeek bool
	ScheduleFrom           string
	ScheduleTo             string
	Service                string
	API                    string
	APIParams              map[string]apiParamStruct
}

type apiParamStruct struct {
	Type      string
	Parameter string
	Content   string
}

type xmlmcResponse struct {
	MethodResult string      `xml:"status,attr"`
	State        stateStruct `xml:"state"`
}

type stateStruct struct {
	Code     string `xml:"code"`
	ErrorRet string `xml:"error"`
}


golang-crosscompile-build.bash
#!/bin/bash
# Orignal https://gist.github.com/jmervine/7d3f455e923cf2ac3c9e
# usage: ./golang-crosscompile-build.bash

#Get current working directory
currentdir=`pwd`

#Clear Sceeen
printf "\033c"

# Get Version out of target then replace . with _
versiond=$(go run *.go -version)
version=${versiond//./_}
#Remove White Space
version=${version// /}
versiond=${versiond// /}
#platforms="darwin/386 darwin/amd64 freebsd/386 freebsd/amd64 freebsd/arm linux/386 linux/amd64 linux/arm windows/386 windows/amd64"
platforms="windows/386 windows/amd64 linux/386 linux/amd64 linux/arm darwin/amd64"
printf " ---- Building API Scheduler $versiond ---- \n"
printf "\n"
for platform in ${platforms}
do
    split=(${platform//\// })
    goos=${split[0]}
    os=${split[0]}
    goarch=${split[1]}
    arch=${split[1]}
    output=goAPIScheduler
    package=goAPIScheduler
    # add exe to windows output
    [[ "windows" == "$goos" ]] && output="$output.exe"
    [[ "windows" == "$goos" ]] && os="win"
    [[ "386" == "$goarch" ]] && arch="x86"
    [[ "amd64" == "$goarch" ]] && arch="x64"

    printf "Platform: $goos - $goarch \n"

    destination="builds/$goos/$goarch/$output"

    printf "Go Build\n"
    GOOS=$goos GOARCH=$goarch go build  -o $destination
    # $target

    printf "Copy Source Files\n"
    #Copy Source to Build Dir
    cp *.md "builds/$goos/$goarch/"
    cp conf*.json "builds/$goos/$goarch/"

    printf "Build Zip \n"
    cd "builds/$goos/$goarch/"
    if [ $os == "darwin" ]; then
        os="osx"
    fi
    zip -r "${package}_${os}_${arch}_v${version}.zip" $output *.md conf*.json > /dev/null
    cp "${package}_${os}_${arch}_v${version}.zip" "../../../${package}_${os}_${arch}_v${version}.zip"
    cd $currentdir
    printf "\n"
done
printf "Clean Up \n"
rm -rf "builds/"
printf "Build Complete \n"
printf "\n"

main.go
package main

import (
	"encoding/json"
	"encoding/xml"
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/go-xmlfmt/xmlfmt"
	"github.com/hornbill/color"
	apiLib "github.com/hornbill/goApiLib"
	"github.com/robfig/cron"
)

func main() {
	//-- Start Time for Log File Name
	timeNow = time.Now().Format("20060102150405")
	//Time for CLI output & Log File Content
	currTime := time.Now().Format(cliTimeLayout)

	//Build map of days of week
	dowMap["SUN"] = 0
	dowMap["MON"] = 1
	dowMap["TUE"] = 2
	dowMap["WED"] = 3
	dowMap["THU"] = 4
	dowMap["FRI"] = 5
	dowMap["SAT"] = 6

	flag.StringVar(&configFileName, "file", "conf.json", "Name of the configuration file to load")
	flag.BoolVar(&configDebug, "debug", false, "Set to true to run scheduler in debug mode, where API call request and reponse XML payload will be written to the log")
	flag.BoolVar(&configDryRun, "dryrun", false, "Set to true to run scheduler in dryrun mode. No API calls will be made")
	flag.BoolVar(&configVersion, "version", false, "Set to true to output the version")
	flag.Parse()
	if configDryRun {
		logEntryType = 6
		logger(logEntryType, "RUNNING IN DRYRUN MODE", true)
	}
	//-- If configVersion just output version number and die
	if configVersion {
		fmt.Printf("%v \n", version)
		return
	}
	logger(logEntryType, "---- Hornbill API Scheduler V"+fmt.Sprintf("%v", version)+" ----", true)
	logger(logEntryType, "Flag - Configuration File: "+configFileName, true)
	logger(logEntryType, "Scheduler started at: "+currTime, true)

	//-- Load Configuration File Into Struct
	apiCallConfig = loadConfig()

	logger(logEntryType, "Enabled Schedule Items:", true)
	c := cron.New()
	c.Start()
	scheduleCount := 0
	for _, scheduleEntry := range apiCallConfig.Schedule {
		scheduleEntryUnique := scheduleEntry
		if scheduleEntryUnique.Enabled {
			scheduleCount++
			logger(logEntryType, "----====----", true)
			logger(logEntryType, "Cron Schedule: "+scheduleEntryUnique.CronSchedule, true)
			logger(logEntryType, "Service: "+scheduleEntryUnique.Service, true)
			logger(logEntryType, "API: "+scheduleEntryUnique.API, true)
			_, err := cron.Parse(scheduleEntryUnique.CronSchedule)
			if err != nil {
				logger(4, "Could not Parse CronSchedule value ["+scheduleEntryUnique.CronSchedule+"] : "+fmt.Sprintf("%v", err), true)
			} else {
				c.AddFunc(scheduleEntryUnique.CronSchedule, func() { apiRequest(scheduleEntryUnique) })
			}
		}
	}

	if scheduleCount == 0 {
		logger(logEntryType, "There are no scheduled tasks enabled!", true)
		return
	}
	logger(logEntryType, "----====----", true)

	//Leave running, wait for user interruption
	sig := make(chan os.Signal)
	signal.Notify(sig, os.Interrupt, os.Kill)
	<-sig

}

func checkSchedule(schedule string) bool {
	boolDOW := false
	boolDOM := false

	schedule = strings.Trim(schedule, " ")
	schedArr := strings.Split(schedule, " ")
	DOW := schedArr[len(schedArr)-1]
	DOM := schedArr[len(schedArr)-3]

	//Get current time and dates
	currDOWNum := strconv.Itoa(int(time.Now().Weekday()))
	currDOWStr := strings.ToUpper(time.Now().Format("Mon"))
	currDOMNum := strings.TrimLeft(time.Now().Format("02"), "0")

	//Check Day Of Week
	//--Check if DOW is wildcard or ignore
	if DOW == "*" || DOW == "?" {
		boolDOW = true
	} else {

		//--Check if DOW is list or single value
		listArr := strings.Split(DOW, ",")
		for _, v := range listArr {
			if v == currDOWNum || strings.ToUpper(v) == currDOWStr {
				boolDOW = true
			}
		}

		//Check if DOW is range
		listArr = strings.Split(DOW, "-")
		if len(listArr) == 2 {
			//Are values int (0-6) or string (SUN-SAT)
			min, err1 := strconv.Atoi(listArr[0])
			max, err2 := strconv.Atoi(listArr[1])
			if err1 != nil && err2 != nil {
				//Couldn't convert, assume range is string SUN to SAT type
				min = dowMap[strings.ToUpper(listArr[0])]
				max = dowMap[strings.ToUpper(listArr[1])]
			}
			intCurrDOWNum, _ := strconv.Atoi(currDOWNum)
			if intCurrDOWNum >= min && intCurrDOWNum <= max {
				boolDOW = true
			}
		}
	}

	//Check Day Of Month
	//--Check if DOM is wildcard or ignore
	if DOM == "*" || DOM == "?" {
		boolDOM = true
	} else {

		//--Check if DOM is list or single value
		listArr := strings.Split(DOM, ",")
		for _, v := range listArr {
			if v == currDOMNum {
				boolDOM = true
			}
		}

		//Check if DOM is range
		listArr = strings.Split(DOM, "-")
		if len(listArr) == 2 {
			min, _ := strconv.Atoi(listArr[0])
			max, _ := strconv.Atoi(listArr[1])
			intCurrDOMNum, _ := strconv.Atoi(currDOMNum)
			if intCurrDOMNum >= min && intCurrDOMNum <= max {
				boolDOM = true
			}
		}
	}

	return (boolDOW && boolDOM)
}

func apiRequest(scheduleEntry apiSchedStruct) {
	if scheduleEntry.DayOfMonthANDDayOfWeek {
		ok := checkSchedule(scheduleEntry.CronSchedule)
		if !ok {
			return
		}
	}
	espXmlmc := apiLib.NewXmlmcInstance(apiCallConfig.InstanceID)
	espXmlmc.SetAPIKey(apiCallConfig.APIKey)

	timeNow := time.Now()
	timeFrom := timeNow
	timeTo := timeNow

	if scheduleEntry.ScheduleFrom != "" || scheduleEntry.ScheduleTo != "" {
		if scheduleEntry.ScheduleFrom != "" {
			timeFromParsed, err := time.Parse(timeLayout, scheduleEntry.ScheduleFrom)
			if err != nil {
				color.Red("\n[ERROR] " + time.Now().Format(cliTimeLayout) + " ScheduleFrom date.time parse failed, unable to run API [" + scheduleEntry.Service + "::" + scheduleEntry.API + "]")
				logger(4, "Unable to parse time string in ScheduleFrom for request ["+scheduleEntry.Service+"::"+scheduleEntry.API+"] : "+fmt.Sprintf("%v", err), false)
				return
			}
			timeFrom = timeFromParsed
		}
		if scheduleEntry.ScheduleTo != "" {
			timeToParsed, err := time.Parse(timeLayout, scheduleEntry.ScheduleTo)
			if err != nil {
				color.Red("\n[ERROR] " + time.Now().Format(cliTimeLayout) + " ScheduleTo date.time parse failed, unable to run API [" + scheduleEntry.Service + "::" + scheduleEntry.API + "]")
				logger(4, "Unable to parse time string in ScheduleTo for request ["+scheduleEntry.Service+"::"+scheduleEntry.API+"] : "+fmt.Sprintf("%v", err), false)
				return
			}
			timeTo = timeToParsed
		}
	}
	if timeTo.Before(timeNow) {
		logger(1, "Current time is after the schedule entry is set to end: "+scheduleEntry.Service+"::"+scheduleEntry.API, true)
		return
	}
	if timeFrom.After(timeNow) {
		logger(1, "Current time is before the schedule entry is set to start: "+scheduleEntry.Service+"::"+scheduleEntry.API, true)
		return
	}

	//Build params
	for i := 0; i < len(scheduleEntry.APIParams); i++ {
		switch scheduleEntry.APIParams[strconv.Itoa(i)].Type {
		case "Content":
			var strContent = scheduleEntry.APIParams[strconv.Itoa(i)].Content

			//Check to see if param content contains nowPlus::int::
			var timeAddRegex = regexp.MustCompile(`nowPlus::[0-9]{0,5}::`)
			var boolDynDate = timeAddRegex.MatchString(strContent)

			//If we have a dynamic date requirement in the config, process
			if boolDynDate {
				result := strings.Split(strContent, "::")
				//Setup time object
				var timeThen = time.Now()
				//Get vals from array
				var timeAddValue, _ = strconv.Atoi(result[1])
				var timeAddUnit = strings.ToLower(result[2])

				if timeAddUnit == "years" {
					timeThen = timeThen.AddDate(timeAddValue, 0, 0)
					strContent = timeThen.Format("2006-01-02 15:04:05")
				} else if timeAddUnit == "months" {
					timeThen = timeThen.AddDate(0, timeAddValue, 0)
					strContent = timeThen.Format("2006-01-02 15:04:05")
				} else if timeAddUnit == "days" {
					timeThen = timeThen.AddDate(0, 0, timeAddValue)
					strContent = timeThen.Format("2006-01-02 15:04:05")
				} else if timeAddUnit == "hours" {
					timeThen = timeThen.Add(time.Duration(timeAddValue) * time.Hour)
					strContent = timeThen.Format("2006-01-02 15:04:05")
				} else if timeAddUnit == "minutes" {
					timeThen = timeThen.Add(time.Duration(timeAddValue) * time.Minute)
					strContent = timeThen.Format("2006-01-02 15:04:05")
				}
			}

			espXmlmc.SetParam(scheduleEntry.APIParams[strconv.Itoa(i)].Parameter, strContent)
		case "Open":
			//Open XML element for complex params
			espXmlmc.OpenElement(scheduleEntry.APIParams[strconv.Itoa(i)].Parameter)
		case "Close":
			//Close XML element for complex params
			espXmlmc.CloseElement(scheduleEntry.APIParams[strconv.Itoa(i)].Parameter)
		}
	}

	var XMLSTRING = espXmlmc.GetParam()
	if configDryRun {
		color.Yellow(time.Now().Format(cliTimeLayout) + " DRYRUN API Call details have been added to the log")
		logger(logEntryType, "----DRYRUN API Request----", false)
		logger(logEntryType, "Service: "+scheduleEntry.Service, false)
		logger(logEntryType, "Method: "+scheduleEntry.API, false)
		logger(logEntryType, "XML Parameters: "+XMLSTRING, false)
		return
	}

	color.Green("\n" + time.Now().Format(cliTimeLayout) + " Executing API Call: " + scheduleEntry.Service + "::" + scheduleEntry.API)
	logger(logEntryType, "Executing API Call: "+scheduleEntry.Service+"::"+scheduleEntry.API, false)
	if configDebug {
		logger(1, "API Request Payload:\n"+xmlfmt.FormatXML(XMLSTRING, "\t", "	"), false)
	}

	XMLAPICall, xmlmcErr := espXmlmc.Invoke(scheduleEntry.Service, scheduleEntry.API)
	if xmlmcErr != nil {
		color.Red(time.Now().Format(cliTimeLayout) + " API Call Failed: " + fmt.Sprintf("%v", xmlmcErr))
		logger(4, "API Call Failed: "+fmt.Sprintf("%v", xmlmcErr), false)
	} else {
		if configDebug {
			logger(1, "API Response Payload:\n"+xmlfmt.FormatXML(XMLAPICall, "\t", "	"), false)
		}

		var xmlRespon xmlmcResponse
		err := xml.Unmarshal([]byte(XMLAPICall), &xmlRespon)
		if err != nil {
			color.Red(time.Now().Format(cliTimeLayout) + " API Call Failed: " + fmt.Sprintf("%v", err))
			logger(4, "API Call Failed: "+fmt.Sprintf("%v", err), false)
		} else {
			if xmlRespon.MethodResult != "ok" {
				color.Red(time.Now().Format(cliTimeLayout) + " API Call Failed: " + xmlRespon.State.ErrorRet)
				logger(4, "API Call Failed: "+xmlRespon.State.ErrorRet, false)
			} else {
				color.Green(time.Now().Format(cliTimeLayout) + " API Call Success: " + scheduleEntry.Service + "::" + scheduleEntry.API)
				logger(logEntryType, "API Call Success: "+scheduleEntry.Service+"::"+scheduleEntry.API, false)
			}
		}
	}
}

//loadConfig -- Function to Load Configruation File
func loadConfig() apiCallStruct {
	//-- Check Config File File Exists
	cwd, _ := os.Getwd()
	configurationFilePath := cwd + "/" + configFileName
	logger(logEntryType, "Loading Config File: "+configurationFilePath, false)
	if _, fileCheckErr := os.Stat(configurationFilePath); os.IsNotExist(fileCheckErr) {
		logger(4, "No Configuration File", true)
		os.Exit(102)
	}
	//-- Load Config File
	file, fileError := os.Open(configurationFilePath)
	//-- Check For Error Reading File
	if fileError != nil {
		logger(4, "Error Opening Configuration File: "+fmt.Sprintf("%v", fileError), true)
		os.Exit(102)
	}

	//-- New Decoder
	decoder := json.NewDecoder(file)
	//-- New Var based on apiCallStruct
	edbConf := apiCallStruct{}
	//-- Decode JSON
	err := decoder.Decode(&edbConf)
	//-- Error Checking
	if err != nil {
		logger(4, "Error Decoding Configuration File: "+fmt.Sprintf("%v", err), true)
		os.Exit(102)
	}
	//-- Return New Config
	return edbConf
}

// logger -- function to append to the current log file
func logger(t int, s string, outputtoCLI bool) {
	cwd, _ := os.Getwd()
	logPath := cwd + "/log"
	logFileName := logPath + "/HB_API_Scheduler_" + timeNow + ".log"

	//-- If Folder Does Not Exist then create it
	if _, err := os.Stat(logPath); os.IsNotExist(err) {
		err := os.Mkdir(logPath, 0777)
		if err != nil {
			color.Red("Error Creating Log Folder %q: %s \r", logPath, err)
			os.Exit(101)
		}
	}

	//-- Open Log File
	f, err := os.OpenFile(logFileName, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0777)
	if err != nil {
		color.Red("Error Creating Log File %q: %s \n", logFileName, err)
		os.Exit(100)
	}
	defer f.Close()
	// assign it to the standard logger
	log.SetOutput(f)
	var errorLogPrefix string
	//-- Create Log Entry
	switch t {
	case 1:
		errorLogPrefix = "[DEBUG] "
		if outputtoCLI {
			color.Set(color.FgGreen)
			defer color.Unset()
		}
	case 2:
		errorLogPrefix = "[MESSAGE] "
		if outputtoCLI {
			color.Set(color.FgGreen)
			defer color.Unset()
		}
	case 3:
		if outputtoCLI {
			color.Set(color.FgGreen)
			defer color.Unset()
		}
	case 4:
		errorLogPrefix = "[ERROR] "
		if outputtoCLI {
			color.Set(color.FgRed)
			defer color.Unset()
		}
	case 5:
		errorLogPrefix = "[WARNING]"
		if outputtoCLI {
			color.Set(color.FgYellow)
			defer color.Unset()
		}
	case 6:
		if outputtoCLI {
			color.Set(color.FgYellow)
			defer color.Unset()
		}
	}

	if outputtoCLI {
		fmt.Printf("%v \n", errorLogPrefix+s)
	}

	log.Println(errorLogPrefix + s)
}
